plugins/basic/python/tk_3dsmaxplus_basic/plugin_bootstrap.py

```diff
@@ -138,14 +126,117 @@ def handle_bootstrap_failed(phase, exception):
+    logger = sgtk.LogManager.get_logger(PLUGIN_PACKAGE_NAME)
     engine = sgtk.platform.current_engine()
+
     if engine:
+        logger.info("Stopping the Shotgun engine.")
+        # Close the various windows (dialogs, panels, etc.) opened by the engine.
+        engine.close_windows()
+        # Turn off your engine! Step away from the car!
         engine.destroy()
+    else:
+        logger.debug("The Shotgun engine was already stopped!")
+
+
+def _on_logout():
+    """
+    Logs the user out and displays login menu
+    """
+
+    _logout_user()
+    _create_login_menu()
+
+
+def _logout_user():
+    """
+    Shuts down the engine and logs out the user of Shotgun.
+    """
+    import sgtk
+
+    # Shutting down the engine also get rid of the engine menu.
+    shutdown_toolkit()
+
+    # Clear the user's credentials to log him/her out.
+    sgtk.authentication.ShotgunAuthenticator().clear_default_user()
+
+
+def _login_user():
+    """
+    Logs in the user to Shotgun and starts the engine.
+    """
+    import sgtk
+    sgtk_logger = sgtk.LogManager.get_logger("PLUGIN_PACKAGE_NAME")
+
+    try:
+        # When the user is not yet authenticated,
+        # pop up the Shotgun login dialog to get the user's credentials,
+        # otherwise, get the cached user's credentials.
+        user = sgtk.authentication.ShotgunAuthenticator().get_user()
+
+    except sgtk.authentication.AuthenticationCancelled:
+        # When the user cancelled the Shotgun login dialog,
+        # keep around the displayed login menu.
+        sgtk_logger.info("Shotgun login was cancelled by the user.")
+        return
+
+    _delete_login_menu()
+
+    # Create a boostrap manager for the logged in user with the plug-in configuration data.
+    toolkit_mgr = sgtk.bootstrap.ToolkitManager(user)
+    toolkit_mgr.base_configuration = constants.BASE_CONFIGURATION
+    toolkit_mgr.plugin_id = constants.PLUGIN_ID
+    toolkit_mgr.bundle_cache_fallback_paths = [os.path.join(plugin_root_path, "bundle_cache")]
+
+    # Retrieve the Shotgun entity type and id when they exist in the environment.
+    # these are passed down through the app launcher when running in zero config
+    entity = toolkit_mgr.get_entity_from_environment()
+    sgtk_logger.debug("Will launch the engine with entity: %s" % entity)
+
+    # set up a simple progress reporter
+    toolkit_mgr.progress_callback = progress_callback
+
+    # start engine
+    sgtk_logger.info("Starting the 3dsmaxplus engine.")
+    toolkit_mgr.bootstrap_engine_async(
+        "tk-3dsmaxplus",
+        entity,
+        completed_callback=handle_bootstrap_completed,
+        failed_callback=handle_bootstrap_failed
+    )
+
+
+def _create_login_menu():
+    """
+    Creates and displays a Shotgun user login menu.
+    """
+
+    _delete_login_menu()
+
+    mb = MaxPlus.MenuBuilder(MENU_LABEL)
+    action = MaxPlus.ActionFactory.Create(
+        'sgtk_plugin_category', "Log In to Shotgun", _login_user)
```

Comment:
Could we also include web links, like maya? https://github.com/shotgunsoftware/tk-maya/blob/develop/zero_config/plugins/basic/python/tk_maya_basic/plugin_logic.py#L250